### ISSUE TITLE
Consider Ornith-1.5-35B-A3B as another model to support

### ISSUE THREAD

**Issue by plavno-plavno** [Neroued/ninfer#81]
<html><head></head><body><h1>Checkpoint proposal: Ornith-1.5-35B-A3B</h1>
<p><strong>Type:</strong> feature request / checkpoint registration
<strong>Target:</strong> third registered NInfer artifact, RTX 5090 (<code>sm_120a</code>)</p>
<hr>
<h2>Why this one</h2>
<p>I understand NInfer is deliberately a closed-set runtime rather than a general
model server, so this is a proposal for one specific addition, not a request to
open up the loader.</p>
<p><a href="https://huggingface.co/ornith-ai/Ornith-1.5-35B-A3B"><code>ornith-ai/Ornith-1.5-35B-A3B</code></a>
(MIT) is an RL-post-trained derivative of Qwen3.5-MoE aimed at agentic coding and
tool use. The relevant point for NInfer is that it should sit almost entirely
inside the code path you already wrote for <code>Qwen3.6-35B-A3B</code>:</p>

Property | Qwen3.6-35B-A3B (registered) | Ornith-1.5-35B-A3B
-- | -- | --
Architecture class | Qwen3_5MoeForConditionalGeneration | Qwen3_5MoeForConditionalGeneration
Topology | hybrid GDN linear attention + softmax attention + sparse MoE | same
Total / active params | 35B / ~3B | 35.1B / ~3B
Native context | 262,144 | 262,144
Multimodal | yes | yes (image-text-to-text)
MTP heads | yes | yes — 785 tensors, separable
License | Apache-2.0 | MIT

<p>Qwen3.6 kept the Qwen3.5 generation-wide architecture; NVIDIA's Megatron Bridge
documents Qwen3.6 as sharing the Qwen3.5 VL MoE implementation and being handled
by the same weight-mapping path. So the delta here is a different <em>fine-tune</em> of
an architecture NInfer already has hand-written kernels for — not a new
architecture.</p>
<h2>Why it's a good fit for NInfer's positioning specifically</h2>
<p>Your published fixtures already include a structured-output and tool-call
scenario, and the MTP acceptance numbers are the headline result. Ornith 1.5 is
post-trained specifically for plan → execute → self-verify agentic loops, which
is the workload where a single-sequence, single-GPU, low-latency engine is most
useful — one developer, one long agent trajectory, no batching needed. It
complements the two general-purpose Qwen checkpoints rather than duplicating
them.</p>
<p>It also fits the memory envelope. BF16 source is ~67 GB; a community NVFP4
weight-only quantization lands at 23 GB on disk, which is in the same range as
your 20.84 GiB <code>qwen3_6_35b_a3b.ninfer</code> artifact.</p>
<h2>Prior art that de-risks bring-up</h2>
<p><a href="https://huggingface.co/r0b0tlab/Ornith-1.5-35B-A3B-NVFP4-W4A16"><code>r0b0tlab/Ornith-1.5-35B-A3B-NVFP4-W4A16</code></a>
is a documented third-party quantization (ModelOpt <code>w4_nvfp4</code> preset, FP8-E4M3 KV
cast) with reproduction scripts and audit logs. Useful findings already
root-caused there:</p>
<ul>
<li>The MTP head is <strong>byte-identically re-attachable</strong> from the source checkpoint
(785 BF16 tensors, SHA-256 audited) — it survives quantization of the trunk.</li>
<li>Speculative decoding works under EAGLE with the draft architecture remapped to
<code>Qwen3_5ForCausalLMMTP</code>; <code>FROZEN_KV_MTP</code> is <em>not</em> implemented for the qwen3_5
MTP head.</li>
<li>Reported acceptance is 0.737 with mean accept length 1.737 at K=1 draft — well
below your 87.2% / 3.62 tokens-per-round at MTP3, but that is a K=1 vs MTP3
comparison, so the ceiling is untested.</li>
<li>Per-tensor NVFP4 reconstruction cosines: experts 0.9956–0.9973, but dense and
GDN projections only 0.967–0.986 under plain max calibration. If NInfer's
conversion recipe uses per-tensor max calibration for the GDN projections, this
is the place to look first.</li>
</ul>
<p>That work was validated on DGX Spark GB10 / SM 12.1, not on <code>sm_120a</code>, so the
runtime specifics (MoE runner backend, FP4 KV availability) won't transfer
directly — but the quantization sensitivity data should.</p>
<h2>Open questions I haven't verified</h2>
<ol>
<li><strong><code>config.json</code> delta</strong> against <code>Qwen/Qwen3.6-35B-A3B</code>: layer count, hidden
size, GDN-to-softmax layer interleave ratio, <code>num_experts</code> /
<code>num_experts_per_tok</code>, <code>head_dim</code>, vocab size. If these match, this is a
weight-mapping job. If they differ, the CUDA-graph and chunked-prefill shape
specialization presumably needs retuning rather than rewriting — you'd know
the real cost there better than I would.</li>
<li><strong>MTP head provenance.</strong> The heads exist and are separable, but Ornith's RL
loop may have left them stale relative to the post-trained trunk. Worth
measuring acceptance at MTP3 before committing.</li>
<li>Whether your conversion tooling under <code>tools/</code> generalizes or is written
against the two current checkpoints' configs.</li>
</ol>

---

## Why this one

I understand NInfer is deliberately a closed-set runtime rather than a general
model server, so this is a proposal for one specific addition, not a request to
open up the loader.

[`[ornith-ai/Ornith-1.5-35B-A3B](https://huggingface.co/ornith-ai/Ornith-1.5-35B-A3B)`](https://huggingface.co/ornith-ai/Ornith-1.5-35B-A3B)
(MIT) is an RL-post-trained derivative of Qwen3.5-MoE aimed at agentic coding and
tool use. The relevant point for NInfer is that it should sit almost entirely
inside the code path you already wrote for `Qwen3.6-35B-A3B`:

| Property | Qwen3.6-35B-A3B (registered) | Ornith-1.5-35B-A3B |
| --- | --- | --- |
| Architecture class | `Qwen3_5MoeForConditionalGeneration` | `Qwen3_5MoeForConditionalGeneration` |
| Topology | hybrid GDN linear attention + softmax attention + sparse MoE | same |
| Total / active params | 35B / ~3B | 35.1B / ~3B |
| Native context | 262,144 | 262,144 |
| Multimodal | yes | yes (image-text-to-text) |
| MTP heads | yes | yes — 785 tensors, separable |
| License | Apache-2.0 | MIT |

Qwen3.6 kept the Qwen3.5 generation-wide architecture; NVIDIA's Megatron Bridge
documents Qwen3.6 as sharing the Qwen3.5 VL MoE implementation and being handled
by the same weight-mapping path. So the delta here is a different *fine-tune* of
an architecture NInfer already has hand-written kernels for — not a new
architecture.

## Why it's a good fit for NInfer's positioning specifically

Your published fixtures already include a structured-output and tool-call
scenario, and the MTP acceptance numbers are the headline result. Ornith 1.5 is
post-trained specifically for plan → execute → self-verify agentic loops, which
is the workload where a single-sequence, single-GPU, low-latency engine is most
useful — one developer, one long agent trajectory, no batching needed. It
complements the two general-purpose Qwen checkpoints rather than duplicating
them.

It also fits the memory envelope. BF16 source is ~67 GB; a community NVFP4
weight-only quantization lands at 23 GB on disk, which is in the same range as
your 20.84 GiB `qwen3_6_35b_a3b.ninfer` artifact.

## Prior art that de-risks bring-up

[`[r0b0tlab/Ornith-1.5-35B-A3B-NVFP4-W4A16](https://huggingface.co/r0b0tlab/Ornith-1.5-35B-A3B-NVFP4-W4A16)`](https://huggingface.co/r0b0tlab/Ornith-1.5-35B-A3B-NVFP4-W4A16)
is a documented third-party quantization (ModelOpt `w4_nvfp4` preset, FP8-E4M3 KV
cast) with reproduction scripts and audit logs. Useful findings already
root-caused there:

- The MTP head is **byte-identically re-attachable** from the source checkpoint
  (785 BF16 tensors, SHA-256 audited) — it survives quantization of the trunk.
- Speculative decoding works under EAGLE with the draft architecture remapped to
  `Qwen3_5ForCausalLMMTP`; `FROZEN_KV_MTP` is *not* implemented for the qwen3_5
  MTP head.
- Reported acceptance is 0.737 with mean accept length 1.737 at K=1 draft — well
  below your 87.2% / 3.62 tokens-per-round at MTP3, but that is a K=1 vs MTP3
  comparison, so the ceiling is untested.
- Per-tensor NVFP4 reconstruction cosines: experts 0.9956–0.9973, but dense and
  GDN projections only 0.967–0.986 under plain max calibration. If NInfer's
  conversion recipe uses per-tensor max calibration for the GDN projections, this
  is the place to look first.

That work was validated on DGX Spark GB10 / SM 12.1, not on `sm_120a`, so the
runtime specifics (MoE runner backend, FP4 KV availability) won't transfer
directly — but the quantization sensitivity data should.

## Open questions I haven't verified

1. **`config.json` delta** against `Qwen/Qwen3.6-35B-A3B`: layer count, hidden
   size, GDN-to-softmax layer interleave ratio, `num_experts` /
   `num_experts_per_tok`, `head_dim`, vocab size. If these match, this is a
   weight-mapping job. If they differ, the CUDA-graph and chunked-prefill shape
   specialization presumably needs retuning rather than rewriting — you'd know
   the real cost there better than I would.
2. **MTP head provenance.** The heads exist and are separable, but Ornith's RL
   loop may have left them stale relative to the post-trained trunk. Worth
   measuring acceptance at MTP3 before committing.
3. Whether your conversion tooling under `tools/` generalizes or is written
   against the two current checkpoints' configs.

**Comment by matthieu-labas** [comment]
Is that AI-generated? A TL;DR would help (because it's definitely TL so I DR...)

**Comment by knoopx** [comment]
it works with no code changes needed, its not a new arch. i have published an ninfer artifact here https://huggingface.co/knoopx/Ornith-1.5-35B-A3B-NInfer with grafted fine tuned mtp head from shisa-ai (the original model ships broken mtp), works fine on my end

**Comment by TomaszDom** [comment]
@knoopx If I'm looking at this correctly, the artifact in https://huggingface.co/huggingJDE/Ornith-1.5-35B-A3B-NInfer is byte-identical?

**Comment by j842** [comment]
Correct I think - mine should be identical. Apologies knoopx and Neroued! I think the issue found in #125 is real though.
